### PR TITLE
How to get consul service discovery for docker

### PR DESCRIPTION
**Pull request by sanglt**
This guide introduce how to get consul working for docker.

After this guide, client can setup their docker service, deploy their image and get the service discovery working.

### CODE REVIEW

**Comment by pbzona** [comment]
Hi @sanglt,

Thanks for your patience, we've been pretty backed up in recent months. This is certainly an interesting concept, and I'll be running some tests over the next few days to make sure everything works as expected. Once I've finished with that, I'll submit a pull request to your branch with any changes and we can go from there. Thanks again, and please let me know if you have questions.

**Comment by pbzona** [comment]
@sanglt Just out of curiosity, which distro did you run this on? I notice the guide you reference as a prerequisite seems to require Ubuntu 14.04, but I'm having trouble getting it to run on 16.04.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+---
+
+*This is a Linode Community guide. [Write for us](/docs/contribute) and earn $250 per published guide.*
+
+<hr>
+
+With the advent of modern [microservices-based architectures](https://medium.com/aws-activate-startup-blog/using-containers-to-build-a-microservices-architecture-6e1b8bacb7d1), many applications are now deployed as a set of distributed components. In such architecture, there is a need to configure and coordinate the various applications running in multiple docker containers across multiple Linode instances
+
+In the following post, I show how a tool called [Consul by HashiCorp](https://www.consul.io) can providing service discovery for a docker. I also provide an example application.
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Tone and persona is inconsistent with style guide.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+## Install Docker
+You should follow the section **Install Docker** in guide [How to install Docker and deploy a LAMP Stack](/docs/applications/containers/how-to-install-docker-and-deploy-a-lamp-stack/)
+Make sure you have docker up and running by running this command and its should have results:
+
+    $ docker info | grep "Server Version"
+    Server Version: 1.11.1
+
+## Getting consul server
+Now its time to get consul server running, with the ui for easy discovery
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Needs a `:` at the end.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    $ docker run -d \
+            --restart always \
+            -p 8500:8500 \
+            --name=consul \
+            progrium/consul \
+            -server \
+            -bootstrap \
+            -ui-dir /ui
+
+{: .note}
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Consider removing note wrapper.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+>`--restart always` Always run this container, even docker has been restart
+>`-p 8500:8500` This binding Consul's HTTP port to Linode address
+>`--name=consul` Set the name of the container
+>`progrium/consul` The name of the docker image to run
+>`-server` Run consul as server mode
+>`-bootstrap` We are starting a consul cluster (of one node) from scratch, rather than joining an existing cluster. We must bootstrap it.
+>`-ui-dir /ui` Enable the consul ui, this docker image have ui source in /ui
+   
+{: .note}
+>In this guide we run its in same server, in production, you may have at least 3 consul server.
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: This sentence is a mess, grammatically speaking.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+## Getting registrator
+Registrator automatically registers and deregisters services for any Docker container by inspecting containers as they come online
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Another missing `:`

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    $ docker run -d \
+         --restart always \
+         --name=registrator \
+         --net=host \
+         --volume=/var/run/docker.sock:/tmp/docker.sock \
+         gliderlabs/registrator \
+           consul://localhost:8500
+
+{: .note}
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Is this not the same as above?

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+You can access the consul ui at http://your_linode_ip:8500/ui, we should have:
+[![consul ui init](/docs/assets/1769-consul-ui-init_small.png)](/docs/assets/1769-consul-ui-init.png)
+
+Or you can check it over REST:
+
+    $ curl http://localhost:8500/v1/catalog/nodes
+      [{"Node":"02fde2c62e87","Address":"172.17.0.5"}]
+
+## Getting some service working
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Lacks title case.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+In real world, and with [microservices-based architectures](https://medium.com/aws-activate-startup-blog/using-containers-to-build-a-microservices-architecture-6e1b8bacb7d1) we will have many of containers running, and we may have 1 service but run in multiple containers (same image).
+In this case, how can we find where is our service.
+
+In this guide I will running 2 services name `order` and `customer` using image `dockercloud/hello-world`. When the container start, docker just map the port avaliable in host to port 80 in container.
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: There's no I in team.

File: docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    $ docker run -d -p 80 -e "NAME=customer01" -e "SERVICE_80_NAME=customer" dockercloud/hello-world
+    $ docker run -d -p 80 -e "NAME=customer02" -e "SERVICE_80_NAME=customer" dockercloud/hello-world
+    
+I use same image here for the example, and this image can say hello to `NAME` in environment, so I use `-e "NAME=order01"`, so that container will tell we know what service + id its running. The `-e "SERVICE_80_NAME=order"` will tell registrator our service at port 80 have name `order`.
+Look at the `-p 80` here, our service is http service and expose at port 80, we can't do the 1-1 mapping here because we may have multiple containers running on same machine like this case. So it just have 1 param, docker will get the available port in host and map it to this port in container.
+
+Now, we should have those services in consul ui:
+[![consul ui service](/docs/assets/1770-consul-ui-services_small.png)](/docs/assets/1770-consul-ui-services.png)
+
+Cool! What about over REST?
```
Comment (alexfornuto) [inline docs/applications/containers/how-to-get-consul-service-discovery-for-docker.md]: Doesn't comply with style guide.

**Comment by alexfornuto** [comment]
I know it's none of my business anymore, but I took a look at this PR and made some notes about where it needs improvement.

**Comment by pbzona** [comment]
Hi @sanglt 

Due to a number of issues discovered while testing this guide, we're going to be unable to publish it. Among other things, the Docker images used in the prerequisite guide are out of date, and without having someone to maintain them, this isn't something we can recommend people follow for a production-ready site. In addition, the prerequisite guide itself hasn't been updated in some time, and we don't want to mix installations across distributions (i.e. it's for Ubuntu 14.04, whereas new guides should use 16.04). As such, we're going to be closing this PR

The topic is very much a good one, however. If you're interested, I can certainly keep you in mind as a contributor for future Docker-related topics that may come up.

**Pull request closed without merging**